### PR TITLE
[struct pack] Detect hash conflict in debug mode

### PR DESCRIPTION
**Pull request by poor-circle**
## Why

Now struct_pack can detected hash conflict in deserialization. 

## What is changing

Add full type info when the NDEBUG marco was not defined. Then we will check it in deserialization.

### CODE REVIEW

**Comment by qicosmos** [comment]
lack of output debug type info?

File: include/coro_rpc/coro_rpc/rpc_protocol.h [hunk 1]
```diff
@@ -68,8 +70,7 @@ using unexpected = tl::unexpected<T>;
 using unexpect_t = tl::unexpect_t;
 #endif
 
-constexpr int SIZE_OF_TYPES_CODE = 4;
-constexpr int RPC_HEAD_LEN = sizeof(rpc_header) + SIZE_OF_TYPES_CODE;
+constexpr int RPC_HEAD_LEN = struct_pack::get_needed_size(rpc_header{});
```
Comment (qicosmos) [inline include/coro_rpc/coro_rpc/rpc_protocol.h:73]: give the exact rpc header len in a comment.

**Comment by qicosmos** [comment]
I think we also need a flag to disable generate debug info at debug mode, because the additional type info it's not good for debugging

File: include/coro_rpc/coro_rpc/rpc_protocol.h (hunk 1 above)
Comment (poor-circle) [reply include/coro_rpc/coro_rpc/rpc_protocol.h:79]: add a static_assert

**Comment by poor-circle** [comment]
> I think we also need a flag to disable generate debug type info at debug mode, because the additional type info it's not good for debugging

Now user can use serialize_config or  enable_type_info to config it.

**Pull request merged**